### ISSUE TITLE
Add P3 and P4 Start

### ISSUE THREAD

**Issue by ToddLa** [yoshisuga/MAME4iOS#220]
If a game has 3 or 4 Players should we add `P3 Start` and `P4 Start` to the in-game menu?

Should we add them to the `HUD` menu? Both?

The reason we did not do this before is..
* the in-game menu was too full/large
* we did not “know for sure” how many players a game had. 

We now have `myosd_num_players` set correctly, and in-game menu has *some* room, so we could easily do this.....

**Comment by ToddLa** [comment]
But the counter argument is if you have a 4 player game, you need 4 game controllers attached, and then you can use the Player Coin and Start from a specific controller. 

That is what you do, for example, in the Simpsons game with 4P

**Comment by rslkr** [comment]
Me and my friends play with Sunset Riders  and Cadillacs and Dinosaurs sometimes.
That's why I asked before for ''Netplay'' for 4 players support. I know that's not your field.

If you saying this causes problem so never mind then.
I don't know how many 3 and 4 player games out there.
But I found this: https://www.recroommasters.com/MAME-4-Player-Games-List-s/76.htm](url)

**Comment by ToddLa** [comment]
It is not a problem, just wondering if useful enough to be in in-game menu. 

Are any of those 4P games turn based, or do you always need 4 controllers?

**Comment by mrjschulte** [comment]
I'm fine with populating the options per the number of players supported by the game.

**Comment by ToddLa** [comment]
Ok I will do it in a separate PR, and rslkr can play with it!

**Comment by rslkr** [comment]
Ah... @ToddLa  You are the No:1 🥇

**Comment by ToddLa** [comment]
Ok as soon as PR#216 goes in I will do another for the P3/4

**Comment by ToddLa** [comment]
There is a branch on my fork called `P3andP4`

**Comment by rslkr** [comment]
**It shows if the game has 3P or 4P**
That's really nice 👍 I guarantee you that no one will complain about the 3P 4P
3P game Dino doesn't show P3 Start on the HUD
I'm guessing b'coz of the order of the rows, no problem its on game menu.
Thank you so much.

**Comment by ToddLa** [comment]
If you see P3 Start on in-game menu, you should also see it on the HUD menu too???

Do the P3 and P4 start do what you expect?

What a Px Start button does is.....

* insert a COIN for P1
* insert a COIN for P1, P2, ...., Px
* push Px START

This works great for 2 Player turn based games.

But for multiplayer games that expect each player to hit start, not so much.

**Comment by ToddLa** [comment]
....maybe what it should do is....

* push Px, ....., P2, P1 START

But I am afraid of what that does to classic 2 player turn based, will they get confused with P2 and P1 START?

**Comment by ToddLa** [comment]
....or maybe push multiple STARTs only in the 3P and 4P case.

**Comment by ToddLa** [comment]
Yes! For 3P games, it does not show on the HUD, the alignent looked weird. It only shows for 4P games. 

What ROM is 3P?

**Comment by ToddLa** [comment]
I guess for 3P games I can show....
                   P1 START
       P2 START P3 START

**Comment by ToddLa** [comment]
....or easier, just show a “blank” for 4P Start

**Comment by rslkr** [comment]
It's not a big problem at all.
It's working on in game menu. 
3P game Cadillacs and Dinosaurs

Another question: when you select P2, P3 or P4 its also insert the coin for P1
This is for turn based games? right
Normally Arcade Machine and desktop MAME every player has own its coin button

**Comment by ToddLa** [comment]
I did a push to my `P3andP4` branch.

* P3 Start will be shown in HUD for only 3P games.
* Replaced redundant `Settings` button in HUD with a `Reset` button (MAME soft reset, aka F3)
* I did *not* push more than one `START` button.

**Comment by ToddLa** [comment]
Yes! we always enter a P1 Coin, (twice!)

* insert a COIN for P1 
* insert a COIN for P1, P2, ...., Px
* push Px START

the extra P1 Coin is “for luck” actually it is for games that require two credits to play by default.

The idea of the P1 and P2 Start buttons in the in-game menu (and now the HUD) is operate as a **DAMN IT!** just start the Eff’n game option!!! So this is why all the extra coins.....

The user still has full control of `COIN` and `START` (a dpad UP+COIN is P2 COIN, or MENU+L1, etc....)

*maybe* for a P3 or P4 Start, we only send a Px COIN and Px START, and keep the current way for P1 and P2

**Comment by ToddLa** [comment]
Did another push, kept the P1,P2 behavior exactly as before.

In the P3+ case:
* insert a COIN for each player.
* hit START for each player.

Also made the joystick OPTION button only insert a COIN and START for the player assigned to the controller (it was also inserting a P1 COIN)

**Comment by ToddLa** [comment]
I am confident I did not change the P1 and P2 path, so we need to test a few P3 and P4 game to make sure my method of pressing all the START buttons is not bad.

But This is still better than the manual SELECT/START on every controller.

And the manual method still works just like before

**Comment by rslkr** [comment]
> ....or easier, just show a “blank” for 4P Start

Yes This better for 3P games.

> But This is still better than the manual SELECT/START on every controller.

Yes, also good for the if controller has fewer buttons or a button stop working, it happened to me before.

> the HUD will use 50% alpha when in Tiny mode.

Maybe no need to add the HUD on game menu, I'm not sure, now it looks better with 50% alpha
Sorry for the inconvenience... ''That's why I said you always come up with better ideas''

And thanks for the ''Reset''  👍 IOY, again.

**Comment by rslkr** [comment]
> I am confident I did not change the P1 and P2 path, so we need to test a few P3 and P4 game to make sure my method of pressing all the START buttons is not bad.

I only checked these 2 games: Sunset Riders and Cadillacs and Dinosaurs
I'll test other games If I find the correct rom for P3/P4
There is still problem with Cadillacs and Dinosaurs!
The blank button activate P1 and P3
P2 Start doesn't start P2 player, for both HUD and in Game Menu
I think doesn't enter coin!!!
I tried with **Joystick right corner+coin** but it works for P1 and P3
L  and Y buttons enter coin start for P2 its in the MAME input/this game.

It seems everything works fine by Sunset Riders.

> the standard COIN, START, OPTIONS, and EXIT on screen buttons are hidden when in fullscreen with the HUD

I like this one as well. Thank you.
